src/hypercorn/app_wrappers.py

```diff
@@ -99,10 +101,48 @@ def start_response(
                 (name.lower().encode("ascii"), value.encode("ascii"))
                 for name, value in response_headers
             ]
+
+        def send_headers(content_length: int) -> None:
+            nonlocal headers, headers_sent
+            if not headers:
+                raise AssertionError("missing call to start_response")
+            if implicit_content_length:  # We can determine the content-length ourself if not set
+                for name, _ in headers:
+                    if name == b"content-length":
+                        break
+                else:  # No content-length set by the application
+                    headers.append((b"content-length", str(content_length).encode("ascii")))
             send({"type": "http.response.start", "status": status_code, "headers": headers})
+            headers_sent = True
+
+        def send_body(data: bytes = b"", more_body: bool = False) -> None:
+            if not headers_sent:
+                send_headers(len(data))
+            send({"type": "http.response.body", "body": data, "more_body": more_body})
+
+        response_body = self.app(environ, start_response)
 
-        for output in self.app(environ, start_response):
-            send({"type": "http.response.body", "body": output, "more_body": True})
+        body_part_count: Optional[int] = None
+        if hasattr(response_body, "__len__"):
+            body_part_count = len(response_body)  # type: ignore
+
+        try:
+            # Optimize the common case of one body part
+            if body_part_count == 1:
+                implicit_content_length = True
+                send_body(next(iter(response_body)))
+            elif body_part_count == 0:
+                send_body()
+            else:
+                for idx, output in enumerate(response_body, start=1):
+                    if output:
+                        more_body = body_part_count is None or idx != body_part_count
+                        send_body(output, more_body=more_body)
+                if body_part_count is None:
```

Comment:
Mhm I just noted that if `response_body` has no item and `body_part_count` is not None we never send a response body. I guess ASGI requires us to at least send a single response body event? This will need tests, noticed as part of https://code.djangoproject.com/ticket/34989#comment:3

Reply:
Nevermind, we explicitly check for length == 0 above, added tests to ensure this behavior is kept.